[client] Handle invalid proxy settings

csordasmarton opened the pull request:
If `HTTP_PROXY` / `HTTPS_PROXY` environment variables are set and the
URL is invalid (e.g.: it doesn't contains a schema like localhost:8001)
the client will throw the following exception:
```
'NoneType' object has no attribute 'rfind'
  File "/codechecker/build/CodeChecker/cc_bin/CodeChecker.py", line 174, in <module>
    main(commands)
  File "/codechecker/build/CodeChecker/cc_bin/CodeChecker.py", line 141, in main
    sys.exit(args.func(args))
  File "//codechecker/build/CodeChecker/lib/python3/codechecker_client/cmd/store.py", line 856, in main
    traceback.print_stack()
```
From this error message it is hard to understand the real problem.

Unfortunately Thrift does not handle the use case when invalid proxy format
is used (e.g.: no schema is specified). For this reason we need to verify
the proxy format in our side.

bruntib commented on web/client/codechecker_client/thrift_call.py:
```diff
@@ -21,6 +21,37 @@
+        if not proxy_settings_are_valid(transport):
+            LOG.error("Invalid proxy format! Check your "
+                      "HTTP_PROXY/HTTPS_PROXY environment variables if "
+                      "these are in the right format:"
```
```suggestion
                      "these are in the right format: "
```

bruntib commented on web/client/codechecker_client/thrift_call.py:
```diff
@@ -21,6 +21,37 @@
 
 LOG = get_logger('system')
 
+PROXY_VALIDATION_NEEDED = True
```
We should consider using a "local static"-ish variable instead of a global one:
```.py
def f():
  f.my_static_var = 42
```

bruntib commented on web/client/codechecker_client/thrift_call.py:
```diff
@@ -21,6 +21,37 @@
+PROXY_VALIDATION_NEEDED = True
+
+
+def proxy_settings_are_valid(transport):
```
This function is not checking the `http[s]` scheme at the beginning, but line 73 says that it will be checked.
Minor thing: maybe the content of this function can go inside `validate_proxy_format()`, since it's quite short.

bruntib commented on web/client/codechecker_client/thrift_call.py:
```diff
@@ -38,6 +69,11 @@ def ThriftClientCall(function):
     funcName = function.__name__
 
     def wrapper(self, *args, **kwargs):
+        # Thrift do not handle the use case when invalid proxy format is used
```
```suggestion
        # Thrift does not handle the use case when invalid proxy format is used
```

csordasmarton commented:
@bruntib I created a new commit where I refactored the whole thing. I moved the checking under the `BaseClientHelper` class. This way we will handle the use case when no port number or invalid port number is set for the proxy environment variables.

zomen2 commented on web/client/codechecker_client/helpers/base.py:
```diff
@@ -28,13 +32,43 @@ def __init__(self, protocol, host, port, uri, session_token=None,
+            # Thrift do not handle the use case when invalid proxy format is
+            # used (e.g.: no schema is specified). For this reason we need to
+            # verify the proxy format in our side.
+            self._validate_proxy_format()
```
If THttpClient throws TTransportException for some reason but not because something bad with proxy settings then this finally branch try to validate proxy settings.
Please call self._validate_proxy_format() in the ValueError branch not in the finally branch.

csordasmarton replied:
Only the `THttpClient` object creation is inside a try/except block, which cannot raise such exception like `TTransportException`.
Two scenarios are possible here:
- The user specified the following environment variable: `HTTP_PROXY=http://localhost` / `HTTP_PROXY=http://localhost:not_a_number` where the port is not specified or it is not an integer. In this case it will raise a ValueError exception.
- The user specified the following environment variable: `HTTP_PROXY=localhost:8001` where the schema is not given. In this case the `THttpClient` object creation will not raise any exception (it will raise exception when an API call happens). So this is the reason when this object creation is done, we need to check the proxy format before any API call happens.

In both cases we need to check the proxy format. So it is not enough to call this function inside the ValueError branch. We need to call it every time when there is no exception.
